nautobot/dcim/tests/test_models.py

```diff
@@ -1155,6 +1158,55 @@ def test_primary_ip_validation_logic(self):
         device.primary_ip6 = interface.ip_addresses.all().filter(ip_version=6).first()
         device.validated_save()
 
+    def test_software_version_device_type_validation(self):
+        """
+        Device's software version must contain a software image that matches the device's device type.
+        """
+
+        software_version = SoftwareVersion.objects.filter(software_images__isnull=False).first()
+
+        self.device_type.software_images.set([])
+        self.device.software_version = software_version
+        with self.assertRaises(ValidationError):
+            self.device.validated_save()
+
+        self.device_type.software_images.add(software_version.software_images.first())
+        self.device.validated_save()
+
+
+class DeviceTypeToSoftwareImageTestCase(ModelTestCases.BaseModelTestCase):
+    model = DeviceTypeToSoftwareImage
+
+    def test_is_default_uniqueness(self):
+        """
+        Assert that only one default software image can be set per device type.
+        """
+
+        device_type = DeviceType.objects.first()
+        software_images = SoftwareImage.objects.all()[:3]
+        DeviceTypeToSoftwareImage.objects.filter(device_type=device_type).delete()
+
+        DeviceTypeToSoftwareImage.objects.create(
+            device_type=device_type,
+            software_image=software_images[0],
+            is_default=True,
+        )
+
+        device_type_to_software_image = DeviceTypeToSoftwareImage(
+            device_type=device_type,
+            software_image=software_images[1],
+            is_default=True,
+        )
+
+        with self.assertRaises(ValidationError):
+            device_type_to_software_image.validated_save()
```

Comment:
Can you point me to where the uniqueness constraint is set here?

Reply:
It's on the DeviceTypeToSoftwareImage model [here](https://github.com/nautobot/nautobot/pull/5137/files#diff-2c74010b539b5fc10880ede3276b50f7af7309db6161e970293b265eebc83a47R116-R123)

Reply:
There's no way to set that property through the ORM from the `DeviceType.software_images` or `SoftwareImage.device_types` today. You have to go directly to the `DeviceTypeToSoftwareImage` model where the validation is already implemented.